### PR TITLE
Software rewrite

### PR DESCRIPTION
**Pull request by KavinSatheeskumar**
This change is [Reviewable](https://reviewable.io/reviews/waterloo-rocketry/rlcs/40)

### CODE REVIEW

**Review by JasonBrave — COMMENTED** [review]
There is some coding style issues to fix

**Review by BluCodeGH — CHANGES_REQUESTED** [review]
Fixed, needs re-testing though.

> Reviewed 63 of 63 files at r1, 13 of 13 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/rlcs/40#-:-NOuOdFj60phR1PUpNpw:bpnae27)* status: all files reviewed, 8 unresolved discussions (waiting on @JasonBrave and @KavinSatheeskumar)

___
*[`src/clientside/config.hpp` line 11 at r1](https://reviewable.io/reviews/waterloo-rocketry/rlcs/40#-NMo1Bw18U35pzNQduUi-r1-11:-NOuHesA-1LoUCUS_kAG:b-896fix) ([raw file](https://github.com/waterloo-rocketry/rlcs/blob/4900c5ffa145250cb5d2371d6bf8e76134668755/src/clientside/config.hpp#L11)):*
<details><summary><i>Previously, JasonBrave (Jason Xu) wrote…</i></summary><blockquote>

I think we can put the constant values in the header file (with `constexpr`)
</blockquote></details>

Done.
___
*[`src/clientside/config.cpp` line 23 at r1](https://reviewable.io/reviews/waterloo-rocketry/rlcs/40#-NMo1Bw18U35pzNQduUh-r1-23:-NOuHvr09zV1oc2O6USd:bt2vr1b) ([raw file](https://github.com/waterloo-rocketry/rlcs/blob/4900c5ffa145250cb5d2371d6bf8e76134668755/src/clientside/config.cpp#L23)):*
<details><summary><i>Previously, JasonBrave (Jason Xu) wrote…</i></summary><blockquote>

Why do we use `uint8_t` type here if we are using `unsigned long` type to store time?
</blockquote></details>

Fixed.
___
*[`src/clientside/lcd.cpp` line 32 at r1](https://reviewable.io/reviews/waterloo-rocketry/rlcs/40#-NMo1Bw18U35pzNQduUl-r1-32:-NOuIqFK6UU5b4RuMH3E:b-896fix) ([raw file](https://github.com/waterloo-rocketry/rlcs/blob/4900c5ffa145250cb5d2371d6bf8e76134668755/src/clientside/lcd.cpp#L32)):*
<details><summary><i>Previously, JasonBrave (Jason Xu) wrote…</i></summary><blockquote>

Find a more appropriate place to put this voltage reading function, it should not be put in LCD display code
</blockquote></details>

Done.
___
*[`src/clientside/lcd.cpp` line 49 at r1](https://reviewable.io/reviews/waterloo-rocketry/rlcs/40#-NMo1Bw18U35pzNQduUl-r1-49:-NOuJHcB708IUVkFsOF1:b-896fix) ([raw file](https://github.com/waterloo-rocketry/rlcs/blob/4900c5ffa145250cb5d2371d6bf8e76134668755/src/clientside/lcd.cpp#L49)):*
<details><summary><i>Previously, JasonBrave (Jason Xu) wrote…</i></summary><blockquote>

Maybe we can display Client-side LiPo voltage while waiting for connection?
</blockquote></details>

Done.
___
*[`src/towerside/config.cpp` line 57 at r1](https://reviewable.io/reviews/waterloo-rocketry/rlcs/40#-NMo1Bw18U35pzNQduV5-r1-57:-NOuLmgdA77_HOpP1Yo_:b-896fix) ([raw file](https://github.com/waterloo-rocketry/rlcs/blob/4900c5ffa145250cb5d2371d6bf8e76134668755/src/towerside/config.cpp#L57)):*
<details><summary><i>Previously, JasonBrave (Jason Xu) wrote…</i></summary><blockquote>

Same as client side timeout, why do we use uint8_t?
</blockquote></details>

Done.
___
*[`src/towerside/errors.cpp` line 3 at r1](https://reviewable.io/reviews/waterloo-rocketry/rlcs/40#-NOuMfnZ7x4fn8ySurAr:-NOuMfnZ7x4fn8ySurAs:bj71bqg) ([raw file](https://github.com/waterloo-rocketry/rlcs/blob/4900c5ffa145250cb5d2371d6bf8e76134668755/src/towerside/errors.cpp#L3)):*
> ```cpp
> #include "errors.hpp"
> 
> namespace errors {
> ```

Just noting that we should do some testing of this errors code, since I don't think it's been explicitly tested.
___
*[`src/towerside/pinout.hpp` line 6 at r1](https://reviewable.io/reviews/waterloo-rocketry/rlcs/40#-NOuMqLZ3eWJq2q-O4JE:-NOuMqLZ3eWJq2q-O4JF:bqtbajs) ([raw file](https://github.com/waterloo-rocketry/rlcs/blob/4900c5ffa145250cb5d2371d6bf8e76134668755/src/towerside/pinout.hpp#L6)):*
> ```cpp
> #include <stdint.h>
> 
> namespace pinout {
> ```

@JasonBrave Does any of this need to change with the new towerside powerboard? Or which serial we initialize in towerside setup()?
___
*[`src/towerside/towerside.ino` line 26 at r1](https://reviewable.io/reviews/waterloo-rocketry/rlcs/40#-NMo1Bw18U35pzNQduVE-r1-26:-NOuOZtM78cu9qbtHr6j:b-896fix) ([raw file](https://github.com/waterloo-rocketry/rlcs/blob/4900c5ffa145250cb5d2371d6bf8e76134668755/src/towerside/towerside.ino#L26)):*
<details><summary><i>Previously, JasonBrave (Jason Xu) wrote…</i></summary><blockquote>

We can remove these debug code
</blockquote></details>

Done.

**Review by KavinSatheeskumar — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/rlcs/40)* status: all files reviewed, 10 unresolved discussions (waiting on @BluCodeGH and @JasonBrave)

___
*[`src/clientside.py` line 6 at r1](https://reviewable.io/reviews/waterloo-rocketry/rlcs/40#-NOMJcey3F6XfD6op4nW:-NOMJcey3F6XfD6op4nX:b-m2rjv7) ([raw file](https://github.com/waterloo-rocketry/rlcs/blob/4900c5ffa145250cb5d2371d6bf8e76134668755/src/clientside.py#L6)):*
> ```Python
> DATA_POINTS = 10
> 
> s = serial.Serial("/dev/ttyUSB0", 115200)
> ```

I don't know if we should be caring all that much about this file, 

but I'd use a with statement, so you can avoid the try, finally



_Code snippet:_
```Python
with serial.Serial("/dev/ttyUSB0", 115200) as s:
    ...
```
___
*[`src/clientside/config.hpp` line 11 at r1](https://reviewable.io/reviews/waterloo-rocketry/rlcs/40#-NMo1Bw18U35pzNQduUi-r1-11:-NOMK2XA783te4l17A87:b-vzbju6) ([raw file](https://github.com/waterloo-rocketry/rlcs/blob/4900c5ffa145250cb5d2371d6bf8e76134668755/src/clientside/config.hpp#L11)):*
<details><summary><i>Previously, JasonBrave (Jason Xu) wrote…</i></summary><blockquote>

I think we can put the constant values in the header file (with `constexpr`)
</blockquote></details>

constexpr is more for things that can be evaluated at compile time (Ie, statements like `x = 3 + 4`). Its main use is for all that template-metaprogramming nonsense. const is the correct thing to use for this context
___
*[`src/clientside/config.hpp` line 13 at r2](https://reviewable.io/reviews/waterloo-rocketry/rlcs/40#-NPA70jyBM2ZTHqeARwD:-NPA70jyBM2ZTHqeARwE:b8rlypj) ([raw file](https://github.com/waterloo-rocketry/rlcs/blob/aca2b00d4b7a5c79f5c9f26dba2ada445a5da024/src/clientside/config.hpp#L13)):*
> ```cpp
> constexpr unsigned long COMMAND_MESSAGE_INTERVAL_MS = 100;
> constexpr unsigned long COMMUNICATION_RESET_MS = 50;
> constexpr uint8_t COMMUNICATION_TIMEOUT_S = 3;
> ```

Did you forget to make this one an unsigned long

**Review by BluCodeGH — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/rlcs/40)* status: all files reviewed, 8 unresolved discussions (waiting on @JasonBrave and @KavinSatheeskumar)

___
*[`src/clientside.py` line 6 at r1](https://reviewable.io/reviews/waterloo-rocketry/rlcs/40#-NOMJcey3F6XfD6op4nW:-NPAV3z18GBjsciEF-8f:bqahsem) ([raw file](https://github.com/waterloo-rocketry/rlcs/blob/4900c5ffa145250cb5d2371d6bf8e76134668755/src/clientside.py#L6)):*
<details><summary><i>Previously, KavinSatheeskumar (Kavin Satheeskumar) wrote…</i></summary><blockquote>

I don't know if we should be caring all that much about this file, 

but I'd use a with statement, so you can avoid the try, finally
</blockquote></details>

Yeah I don't really care about this file, its just a testing script.
___
*[`src/clientside/config.hpp` line 11 at r1](https://reviewable.io/reviews/waterloo-rocketry/rlcs/40#-NMo1Bw18U35pzNQduUi-r1-11:-NPAVAw74CThtynrdpJ9:bksz3ry) ([raw file](https://github.com/waterloo-rocketry/rlcs/blob/4900c5ffa145250cb5d2371d6bf8e76134668755/src/clientside/config.hpp#L11)):*
<details><summary><i>Previously, KavinSatheeskumar (Kavin Satheeskumar) wrote…</i></summary><blockquote>

constexpr is more for things that can be evaluated at compile time (Ie, statements like `x = 3 + 4`). Its main use is for all that template-metaprogramming nonsense. const is the correct thing to use for this context
</blockquote></details>

Constexpr works and makes the code cleaner so I think we should stick with it.
___
*[`src/clientside/config.hpp` line 13 at r2](https://reviewable.io/reviews/waterloo-rocketry/rlcs/40#-NPA70jyBM2ZTHqeARwD:-NPAVE1l4Ql0mtMH_I7x:bba25od) ([raw file](https://github.com/waterloo-rocketry/rlcs/blob/aca2b00d4b7a5c79f5c9f26dba2ada445a5da024/src/clientside/config.hpp#L13)):*
<details><summary><i>Previously, KavinSatheeskumar (Kavin Satheeskumar) wrote…</i></summary><blockquote>

Did you forget to make this one an unsigned long
</blockquote></details>

No, this is a number in seconds and is compared against another uint8_t.

**Review by KavinSatheeskumar — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/rlcs/40#-:-NPAfhHk8F1O9I3TF7yG:bnfp4nl)* status: all files reviewed, 8 unresolved discussions (waiting on @JasonBrave)

**Review by BluCodeGH — APPROVED** [review]
@JasonBrave please go through reviewable and mark things as resolved so we can merge this.

> Reviewed 9 of 9 files at r3, 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/rlcs/40#-:-NQaaU3JESHS3SfRYVLy:bsw9eku)* status: all files reviewed, 7 unresolved discussions (waiting on @JasonBrave and @KavinSatheeskumar)

File: src/common/communication.hpp [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+      buffer_position = 0;
+      return false;
+    }
+
+    memcpy(dest, receive_buffer + 1, sizeof(RT));
+    buffer_position = 0;
+    return true;
+  }
+
+  uint8_t seconds_since_last_contact() {
```
Comment (JasonBrave) [inline src/common/communication.hpp:45]: Potential uint8_t overflow? If lost communication for more than 4 minutes.

**Review by JasonBrave — DISMISSED** [review]

**Pull request merged**